rust/lance/src/index/append.rs

```diff
@@ -56,25 +59,57 @@ pub async fn append_index(
         })?;
 
     let index = dataset
-        .open_vector_index(&column.name, old_index.uuid.to_string().as_str())
+        .open_generic_index(&column.name, &old_index.uuid.to_string())
         .await?;
 
-    let Some(ivf_idx) = index.as_any().downcast_ref::<IVFIndex>() else {
-        info!("Index type: {:?} does not support append", index);
-        return Ok(None);
-    };
-
-    let mut scanner = dataset.scan();
-    scanner.with_fragments(unindexed);
-    scanner.with_row_id();
-    scanner.project(&[&column.name])?;
-    let stream = scanner.try_into_stream().await?;
-
-    let new_index = ivf_idx
-        .append(dataset.as_ref(), stream, old_index, &column.name)
-        .await?;
-
-    Ok(Some((new_index, frag_bitmap)))
+    match index.index_type() {
+        IndexType::Scalar => {
+            let index = dataset
+                .open_scalar_index(&column.name, &old_index.uuid.to_string())
+                .await?;
+
+            let mut scanner = dataset.scan();
+            scanner
+                .with_fragments(unindexed)
+                .with_row_id()
+                .order_by(Some(vec![ColumnOrdering::asc_nulls_first(
+                    column.name.clone(),
+                )]))?
+                .project(&[&column.name])?;
+            let new_data_stream = scanner.try_into_stream().await?;
+
+            let new_uuid = Uuid::new_v4();
+
+            let index_dir = dataset.indices_dir().child(new_uuid.to_string());
+            let new_store = LanceIndexStore::new((*dataset.object_store).clone(), index_dir);
+
+            index.update(new_data_stream.into(), &new_store).await?;
```

Comment:
Remind me, why don't we remove old data? Is that a future TODO?

Reply:
Hmm...embarrassingly I think it's just because I didn't consider it :laughing: I'll add it as a future TODO.